software/bootcamp-code/main.py

```diff
@@ -33,56 +56,170 @@ def main_loop():
+                    y_speed_prop=(processedData.balls[-1].distance-400)/(cam.rgb_height-400)
+                    # rotational speed is the difference between the desired x-location of the ball and actual, normalized and proportional
+                    rot_speed_prop= (ball_desired_x - processedData.balls[-1].x)/cam.rgb_width
+                
+                    # if the basket and ball are in the center of the frame and ball is close enough move on to throwing
+                    if -0.05>x_speed_prop>0.05 and -0.05>y_speed_prop>0.05 and -0.1<y_speed_prop:
+                        state = State.THROW_BALL
+                        continue
+                    # center the basket and the ball with orbiting, get the ball to the desired distance
+                    omni_motion.move(orbit_speed*x_speed_prop, orbit_speed*y_speed_prop, orbit_speed*rot_speed_prop)
+
+
+            # drive ontop of the ball and throw it.
+            elif state==State.THROW_BALL:
+                last_state, new_state = state_printer(state, last_state, new_state)
+                
+                # enters the if statement once to start the throw timer when the ball is out of frame
+                if ball_out_of_frame==False:
+                    if (len(processedData.balls)<1:
+                        ball_out_of_frame=True
+                        throw_start=time.time()
+                    elif processedData.balls[-1].distance<400:
+                        ball_out_of_frame=True
+                        throw_start=time.time()
+                
+                # if the ball is out of frame then the throw duration should be bigger than 1.2
+                throw_duration=time.time()-throw_start
+                if throw_duration>1.2 and ball_out_of_frame==True:
+                    state=State.FIND_BALL
+                    throw_start=0
+                    ball_out_of_frame=False
+
+                
+                elif ball_out_of_frame==True:
+                    #when the ball is not in view, calculate proportional speed for the thrower
+                    omni_motion.move(0, -1*throw_forward_speed, 0, throw_motor_speed)
+                
+                # when the ball is in view, drive towards it
+                else:
+                    #?????? For the thrower motor speeds, I suggest mapping the mainboard-speed to throwing distance. 
+                    # Based on that you can either estimate a function or linearly interpolate the speeds.
+                    
+                    # normalize the ball distance from the robot to the 0-1 range
+                    basket_dist_norm = (basket_max_distance-processedData.basket_b.xbasket_max_distance
+                    # proportional speed to the distance of the ball
+                    throw_motor_speed = throw_motor_speed_max/basket_dist_norm
+                    omni_motion.move(0, -1*throw_forward_speed, 0, throw_motor_speed)
```

Comment:
Is it good to just drive forward? What happens if someting bumps the robot or if the robot just does not drive straight

Reply:
Tried to implement the thrower speed to be proportional to distances. Should the motor speed be proportional to the basket distance?